ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+        msg_tag = hash(uuid.uuid4())
+        ctrl_tag = hash(uuid.uuid4())
+        peer_info = await exchange_peer_info(
+            ucp_endpoint=ucp_endpoint,
+            msg_tag=msg_tag,
+            ctrl_tag=ctrl_tag,
+            guarantee_msg_order=guarantee_msg_order,
+        )
+        ep = public_api.Endpoint(
+            ucp_endpoint=ucp_endpoint,
+            worker=worker,
+            ctx=ctx,
+            msg_tag_send=peer_info["msg_tag"],
+            msg_tag_recv=msg_tag,
+            ctrl_tag_send=peer_info["ctrl_tag"],
+            ctrl_tag_recv=ctrl_tag,
+            guarantee_msg_order=guarantee_msg_order,
+        )
+
+        logging.debug(
+            "listener_handler() server: %s, msg-tag-send: %s, "
+            "msg-tag-recv: %s, ctrl-tag-send: %s, ctrl-tag-recv: %s"
+            % (
+                hex(ucp_endpoint.handle),
+                hex(ep._msg_tag_send),
+                hex(ep._msg_tag_recv),
+                hex(ep._ctrl_tag_send),
+                hex(ep._ctrl_tag_recv),
+            )
+        )
+
+        # Setup the control receive
+        setup_ctrl_recv(ep)
+
+        # Removing references here to avoid delayed clean up
+        del ctx
+
+        # Finally, we call `func` asynchronously (even if it isn't coroutine)
+        if asyncio.iscoroutinefunction(func):
+            await func(ep)
+        else:
+
+            async def _func(ep):  # coroutine wrapper
+                func(ep)
+
+            await _func(ep)
+
+    asyncio.ensure_future(run(ucp_endpoint, ctx, worker, func, guarantee_msg_order))
+
+
+async def _non_blocking_mode(weakref_ctx):
+    """This help function maintains a UCX progress loop.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        del ctx
+        await asyncio.sleep(0)
+
+
+async def _arm_worker(weakref_ctx, rsock, event_loop):
+    """This help function args the worker.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+
+    # When arming the worker, the following must be true:
+    #  - No more progress in UCX (see doc of ucp_worker_arm())
+    #  - All asyncio tasks that isn't waiting on UCX must be executed
+    #    so that the asyncio's next state is epoll wait.
+    #    See <https://github.com/rapidsai/ucx-py/issues/413>
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        # This IO task returns when all non-IO tasks are finished.
+        await event_loop.sock_recv(rsock, 1)
+        if ctx._worker.arm():
+            # At this point we know that asyncio's next state is
+            # epoll wait.
+            break
+        del ctx
```

Comment:
Won't this scope out normally without the `del`? Or is there some reason it is explicitly needed?

Reply:
You are right, fixed